### PR TITLE
updateEpoch need copy trie node, and updateEpoch after recursion succ…

### PR DESCRIPTION
**Pull request by cryyl**
### Description

updateEpoch need copy trie node, and updateEpoch after recursion success call

### Rationale

updateEpoch need copy trie node, and updateEpoch after recursion success call

### Example

None

### Changes

Notable changes: 
* updateEpoch need copy trie node, and updateEpoch after recursion success call

### CODE REVIEW

File: trie/trie.go [hunk 1]
```diff
@@ -439,11 +447,13 @@ func (t *Trie) insert(n node, prefix, key []byte, value node, epoch types.StateE
 		// If the whole key matches, keep this short node as is
 		// and only update the value.
 		if matchlen == len(n.Key) {
-			n.setEpoch(t.currentEpoch)
 			dirty, nn, err := t.insert(n.Val, append(prefix, key[:matchlen]...), key[matchlen:], value, n.epoch)
 			if !dirty || err != nil {
 				return false, n, err
 			}
+			if t.withShadowNodes {
```
Comment (0xbundler) [inline trie/trie.go:454]: may need not update epoch, it reinit a shortNode with `t.currentEpoch`

File: trie/trie.go [hunk 2]
```diff
@@ -576,6 +585,9 @@ func (t *Trie) delete(n node, prefix, key []byte, epoch types.StateEpoch) (bool,
 		if !dirty || err != nil {
```
Comment (0xbundler) [inline trie/trie.go:585]: when update epoch, it need copy too.

File: trie/trie.go [hunk 3]
```diff
@@ -266,15 +270,19 @@ func (t *Trie) tryGetWithEpoch(origNode node, key []byte, pos int, epoch types.S
...
-		if err == nil && didResolve {
-			n = n.copy()
-			n.Children[key[pos]] = newnode
+		if err == nil {
+			if didResolve || (updateEpoch && epoch != t.currentEpoch) {
+				n = n.copy()
+				n.Children[key[pos]] = newnode
+				if updateEpoch && epoch != t.currentEpoch {
+					n.setEpoch(t.currentEpoch)
+					n.UpdateChildEpoch(int(key[pos]),  t.currentEpoch)
```
Comment (0xbundler) [inline trie/trie.go:283]: when child is nil, may need not update its epoch?

Comment (cryyl) [reply trie/trie.go:283]: if child is nil, err could not be nil

File: trie/trie.go (hunk 2 above)
Comment (cryyl) [reply trie/trie.go:585]: resolve

File: trie/trie.go (hunk 1 above)
Comment (cryyl) [reply trie/trie.go:454]: resolve

File: trie/trie.go [hunk 4]
```diff
@@ -245,14 +245,18 @@ func (t *Trie) tryGetWithEpoch(origNode node, key []byte, pos int, epoch types.S
...
 		if updateEpoch {
-			n.setEpoch(t.currentEpoch)
 			value, newnode, didResolve, err = t.tryGetWithEpoch(n.Val, key, pos+len(n.Key), t.currentEpoch, true)
 		} else {
 			value, newnode, didResolve, err = t.tryGetWithEpoch(n.Val, key, pos+len(n.Key), epoch, false)
 		}
-		if err == nil && didResolve {
-			n = n.copy()
-			n.Val = newnode
+		if err == nil {
```
[the remark this answers is not in the document]
Comment (cryyl) [reply trie/trie.go:252]: resolve

File: trie/trie.go [hunk 5]
```diff
@@ -245,14 +245,18 @@ func (t *Trie) tryGetWithEpoch(origNode node, key []byte, pos int, epoch types.S
 		}
 
 		if updateEpoch {
```
[the remark this answers is not in the document]
Comment (cryyl) [reply trie/trie.go:247]: this two call have different epoch argument

[the remark this answers is not in the document]
Comment (0xbundler) [reply trie/trie.go:247]: I mean just pass updateEpoch is ok, when updateEpoch is true, it will pass true too.

File: trie/trie.go [hunk 6]
```diff
@@ -245,14 +245,16 @@ func (t *Trie) tryGetWithEpoch(origNode node, key []byte, pos int, epoch types.S
...
-			n.setEpoch(t.currentEpoch)
 			value, newnode, didResolve, err = t.tryGetWithEpoch(n.Val, key, pos+len(n.Key), t.currentEpoch, true)
 		} else {
 			value, newnode, didResolve, err = t.tryGetWithEpoch(n.Val, key, pos+len(n.Key), epoch, false)
 		}
-		if err == nil && didResolve {
+		if err == nil && didResolve || (updateEpoch && epoch != t.currentEpoch) {
```
Comment (0xbundler) [inline trie/trie.go:252]: may using `err == nil && (didResolve || (updateEpoch && epoch != t.currentEpoch))` is safe, or using a function

**Pull request closed without merging**